### PR TITLE
V22F-399 create shared object

### PR DESCRIPTION
**Pull request by KungFuDonkey**
Creates a shared object so speed dreams and sdaLib can send data to each other via pointers

### CODE REVIEW

**Review by noutheesink — APPROVED** [review]

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/BlackBoxData.cpp [hunk 1]
```diff
@@ -1,6 +1,871 @@
...
+    Car.priv.smoke = p_priv.smoke;
+    Car.priv.normal.x = p_priv.normal.x;
+    Car.priv.normal.y = p_priv.normal.y;
+    Car.priv.normal.z = p_priv.normal.z;
+    Car.priv.collpos.x = p_priv.collpos.x;
+    Car.priv.collpos.y = p_priv.collpos.y;
+    Car.priv.collpos.z = p_priv.collpos.z;
+    Car.priv.dammage = p_priv.dammage;
+    Car.priv.debug = p_priv.debug;
+    Car.priv.collision_state.collision_count = p_priv.collision_state.collision_count;
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/BlackBoxData.cpp:306]:
```suggestion
    Car.priv.collision    = p_priv.collision;
    Car.priv.simcollision = p_priv.simcollision;
    Car.priv.smoke        = p_priv.smoke;
    
    Car.priv.normal.x = p_priv.normal.x;
    Car.priv.normal.y = p_priv.normal.y;
    Car.priv.normal.z = p_priv.normal.z;
    
    Car.priv.collpos.x = p_priv.collpos.x;
    Car.priv.collpos.y = p_priv.collpos.y;
    Car.priv.collpos.z = p_priv.collpos.z;
    
    Car.priv.dammage = p_priv.dammage;
    Car.priv.debug   = p_priv.debug;
    Car.priv.collision_state.collision_count = p_priv.collision_state.collision_count;
```

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/IPCPointerManager.inl [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include "IPCPointerManager.h"
+
+#define CREATE_IPC_POINTER_MANAGER_IMPLEMENTATION(p_type) \
+    template IPCPointerManager<p_type>::IPCPointerManager(); \
+    template p_type* IPCPointerManager<p_type>::GetDataPointer() const; \
+    template tTrackSeg* IPCPointerManager<p_type>::GetSegmentPointer() const;
+
+
+template <typename BlackBoxData>
+IPCPointerManager<BlackBoxData>::IPCPointerManager()
```
[the remark this answers is not in the document]
Comment (shoekzema) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/IPCPointerManager.inl:12]: +1

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SocketBlackBox.inl [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
-	template void SocketBlackBox<type>::Shutdown();\
-	template void SocketBlackBox<type>::SerializeBlackBoxData(msgpack::sbuffer& p_sbuffer, BlackBoxData* p_blackBoxData);\
-    template void SocketBlackBox<type>::DeserializeBlackBoxResults(const char* p_dataReceived, unsigned int p_size, DecisionTuple& p_decisionTuple);\
-    template bool SocketBlackBox<type>::GetDecisions(tCarElt* p_car, tSituation* p_situation, int p_tickCount, DecisionTuple& p_decisions);
+#define CREATE_SOCKET_BLACKBOX_IMPLEMENTATION(p_type1,p_type2) \
+	template SocketBlackBox<p_type1,p_type2>::SocketBlackBox(PCWSTR p_ip, int p_port);\
+    template void SocketBlackBox<p_type1,p_type2>::Initialize();\
+	template void SocketBlackBox<p_type1,p_type2>::Initialize(BlackBoxData& p_initialBlackBoxData, BlackBoxData* p_tests, int p_amountOfTests);\
+	template void SocketBlackBox<p_type1,p_type2>::Shutdown();\
+	template void SocketBlackBox<p_type1,p_type2>::SerializeBlackBoxData(msgpack::sbuffer& p_sbuffer, BlackBoxData* p_BlackBoxData);\
```
Comment (shoekzema) [inline speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/SocketBlackBox.inl:13]: p_BlackBoxData to p_blackBoxData
```suggestion
	template void SocketBlackBox<p_type1,p_type2>::SerializeBlackBoxData(msgpack::sbuffer& p_sbuffer, BlackBoxData* p_blackBoxData);\
```

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/BlackBoxData.cpp [hunk 4]
```diff
@@ -1,6 +1,871 @@
...
+#include "robot.h"
+
+/**
+ * Constructs a data type for holding data provided to the black box
+ * Makes deep copies of the provided data
+ * @param p_car       The car data in the simulation
+ * @param p_situation The environment data in the simulation
+ * @param p_tickCount The tick that this driving simulation is from
+ */
+BlackBoxData::BlackBoxData(tCarElt* p_car, tSituation* p_situation, int p_tickCount, tTrackSeg* p_nextSegments, int p_nextSegmentsCount) : TickCount(p_tickCount)
```
[the remark this answers is not in the document]
Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/BlackBoxData.cpp:12]: will be fixed in another branch

File: speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/CarController.cpp [hunk 5]
```diff
@@ -13,7 +13,7 @@ void CarController::SetSteerCmd(float p_steer) const
 /// @param  p_accel The amount that needs to be edited
 void CarController::SetAccelCmd(float p_accel) const
 {
-	m_car->ctrl.accelCmd = p_accel;
+	m_car->ctrl.accelCmd += p_accel;
```
[the remark this answers is not in the document]
Comment (KungFuDonkey) [reply speed-dreams/source-2.2.3/src/simulatedDrivingAssistance/backend/CarController.cpp:16]: Then the rest needs to be renamed too, but we will keep this in mind for the next refactor

**Review by larsdekwant — APPROVED** [review]

**Pull request merged**